### PR TITLE
refactor(versioning/generic): Strict null checks

### PR DESCRIPTION
**Pull request by zharinov**
## Changes:

- Enable strict null checks for generic versioning and its remaining users

## Context:

- Ref: #7154

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/versioning/regex/index.ts [hunk 1]
```diff
@@ -66,66 +66,95 @@ export class RegExpVersioningApi extends GenericVersioningApi<RegExpVersion> {
...
       return null;
     }
 
-    const groups = match.groups;
+    const { major, minor, patch, build, prerelease, compatibility } = groups;
     const release = [
-      typeof groups.major === 'undefined' ? 0 : Number(groups.major),
-      typeof groups.minor === 'undefined' ? 0 : Number(groups.minor),
-      typeof groups.patch === 'undefined' ? 0 : Number(groups.patch),
+      typeof major === 'undefined' ? 0 : Number(major),
```
Comment (viceice) [inline lib/versioning/regex/index.ts:76]:
```suggestion
      typeof major === 'undefined' ? 0 : parseInt(major, 10),
```
This is more safe

File: lib/versioning/regex/index.ts [hunk 2]
```diff
@@ -66,66 +66,95 @@ export class RegExpVersioningApi extends GenericVersioningApi<RegExpVersion> {
...
+    const parsedVersion = this._parse(version);
+    const parsedRange = this._parse(range);
+    return !!(
+      parsedVersion &&
+      parsedRange &&
+      semver.ltr(asSemver(parsedVersion), asSemver(parsedRange))
     );
   }
 
+  private getSemverVersions(versions: string[]): string[] {
```
Comment (viceice) [inline lib/versioning/regex/index.ts:112]:
```suggestion
  private _getSemverVersions(versions: string[]): string[] {
```

i think we skip functions with underscore in tests

**Review by rarkins — DISMISSED** [review]

File: tsconfig.strict.json [hunk 3]
```diff
@@ -391,16 +391,15 @@
     "lib/versioning/aws-machine-image/index.ts",
     "lib/versioning/common.ts",
     "lib/versioning/composer/index.ts",
+    "lib/versioning/docker/index.ts",
```
Comment (rarkins) [inline tsconfig.strict.json:394]: Added back?

File: lib/versioning/regex/index.ts [hunk 4]
```diff
@@ -66,66 +66,93 @@ export class RegExpVersioningApi extends GenericVersioningApi<RegExpVersion> {
...
 
-    const groups = match.groups;
+    const { major, minor, patch, build, prerelease, compatibility } = groups;
     const release = [
-      typeof groups.major === 'undefined' ? 0 : Number(groups.major),
-      typeof groups.minor === 'undefined' ? 0 : Number(groups.minor),
-      typeof groups.patch === 'undefined' ? 0 : Number(groups.patch),
+      typeof major === 'undefined' ? 0 : parseInt(major, 10),
+      typeof minor === 'undefined' ? 0 : Number(minor),
+      typeof patch === 'undefined' ? 0 : Number(patch),
```
Comment (viceice) [inline lib/versioning/regex/index.ts:78]: There are some more `parseInt` candidate's

File: lib/versioning/regex/index.ts [hunk 5]
```diff
@@ -66,66 +66,93 @@ export class RegExpVersioningApi extends GenericVersioningApi<RegExpVersion> {
...
     ];
 
-    if (groups.build) {
-      release.push(Number(groups.build));
+    if (build) {
+      release.push(Number(build));
```
Comment (viceice) [inline lib/versioning/regex/index.ts:82]: Here

**Review by rarkins — APPROVED** [review]

**Pull request merged**